# EDA

In [1]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 50)
pd.set_option('display.precision', 4)

In [2]:
from acquiring_xsell.data import load_data

base_df = load_data()

base_df.head()

,month,inn,inn_status,top_mcc_group_inn,who_is_this_first_inn,first_trx_month_inn,last_active_month_inn,lifetime_month_streak_inn,real_kam_on_inn,has_termination_start,altpay1_turnover,altpay2_turnover,altpay3_turnover,altpay4_turnover,altpay5_turnover,altpay6_turnover,has_recurrent_payments,cnt_trx,turnover,other_active_product_cnt,is_relevant_mcc_p3,is_relevant_mcc_altpay5,is_relevant_turnover_altpay5,active_mcc_cnt,active_gateway_cnt,avg_check,turnover_ma_2m,turnover_ema_2m,turnover_ma_3m,turnover_ema_3m,turnover_ma_6m,turnover_ema_6m,cnt_trx_ma_2m,cnt_trx_ema_2m,cnt_trx_ma_3m,cnt_trx_ema_3m,cnt_trx_ma_6m,cnt_trx_ema_6m,avg_check_wma_2m,avg_check_wma_3m,avg_check_wma_6m,turnover_cv_2m,turnover_cv_3m,turnover_cv_6m,turnover_change_to_prev_month,cnt_trx_change_to_prev_month,avg_check_change_to_prev_month,turnover_change_to_turnover_ma_2m,turnover_change_to_turnover_ma_3m,turnover_change_to_turnover_ma_6m,turnover_change_to_turnover_ema_2m,turnover_change_to_turnover_ema_3m,turnover_change_to_turnover_ema_6m,cnt_trx_change_to_cnt_trx_ma_2m,cnt_trx_change_to_cnt_trx_ma_3m,cnt_trx_change_to_cnt_trx_ma_6m,cnt_trx_change_to_cnt_trx_ema_2m,cnt_trx_change_to_cnt_trx_ema_3m,cnt_trx_change_to_cnt_trx_ema_6m,avg_check_change_to_avg_check_wma_2m,avg_check_change_to_avg_check_wma_3m,avg_check_change_to_avg_check_wma_6m,first_month_p1,first_month_p2,first_month_p3,first_month_altpay5
0,2024-12-01 00:00:00.000,INN_14411,Churn,Tickets&clubs,referral,2021-04-01 00:00:00.000,2024-07-01 00:00:00.000,0,False,False,0.0000,0.0,0.0,0.0,0.0,0.0,False,0.0000,0.0000,NaN,False,False,False,0,0,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,NaN,NaN,NaN,NaN,NaN,NaN,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN
1,2025-01-01 00:00:00.000,INN_14411,Churn,Tickets&clubs,referral,2021-04-01 00:00:00.000,2024-07-01 00:00:00.000,0,True,False,0.0000,0.0,0.0,0.0,0.0,0.0,False,0.0000,0.0000,NaN,False,False,False,0,0,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN
2,2024-12-01 00:00:00.000,INN_24790,Churn,Travel,outcoming,2022-07-01 00:00:00.000,2024-08-01 00:00:00.000,0,False,False,0.0000,0.0,0.0,0.0,0.0,0.0,False,0.0000,0.0000,NaN,False,False,False,0,0,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,NaN,NaN,NaN,NaN,NaN,NaN,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN
3,2025-01-01 00:00:00.000,INN_24790,Churn,Travel,outcoming,2022-07-01 00:00:00.000,2024-08-01 00:00:00.000,0,True,False,0.0000,0.0,0.0,0.0,0.0,0.0,False,0.0000,0.0000,NaN,False,False,False,0,0,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN
4,2024-12-01 00:00:00.000,INN_1746,Active,Digital,outcoming,2022-08-01 00:00:00.000,2024-12-01 00:00:00.000,29,False,False,0.0012,0.0,0.0,0.0,0.0,0.0,False,0.0049,0.0064,NaN,False,True,False,2,1,0.0008,0.0066,0.0056,0.0029,0.0036,0.0085,0.0055,0.0062,0.0205,0.0057,0.0142,0.0029,0.0123,0.0012,0.0009,0.0015,NaN,NaN,NaN,NaN,NaN,NaN,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN


## Описание датасета
Иcходный датасет содержат данные по месяцам жизни каждой компании-клиента интернет-эквайринга за период декабрь 2024 — февраль 2026.

Условные обозначения:
- `W` — размер окна скользящего среднего. Принимает значения `[2, 3, 6]`
- `Y` — дополнительные продукты `['p1', 'p2', 'p3']` и платёжный метод `'altpay_5'`
- `N` — платёжные методы `[1, 2, 3, 4, 5]`

### Категориальные переменные
#### Порядковые
- `month` — календарный месяц
- `first_trx_month_inn` — месяц, в котором у компании была первая боевая (т.е. не тестовая) транзакция. В интернет-эквайринге считается первым месяцем жизни клиента. Константный признак в рамках одной ИНН
- `last_active_month_inn` — последний месяц активности компании на момент текущего месяца `month`
- `first_month_Y` — месяц, в котором у ИНН впервые появился продукт `Y`. **Константный признак в рамках одной ИНН**
#### Номинальные
- `inn` — уникальный идентификатор компании. Константный признак в рамках одной ИНН
- `inn_status` — статус оттока компании на момент текущего месяца `month`. Принимает значения `['Active', 'New Churn', 'Churn', 'Reborn']`
- `top_mcc_group_inn` — группа MCC (merchant category code), по которой у данной компании был наибольший оборот в текущем месяце `month`
- `who_is_this_first_inn` — канал привлечения компании. Принимает значения: `['incoming', 'outcoming', 'referral']`. Константный признак в рамках одной ИНН
#### Бинарные
- `real_kam_on_inn` — наличие закреплённого Key Account Manager у компании на момент текущего месяца `month`
- `has_termination_start` — находится ли по данным ФНС компания в процессе ликвидации на момент текущего месяца `month`
- `has_recurrent_payments` — были ли рекуррентные платежи у компании на момент текущего месяца `month`
- `is_relevant_mcc_p3` — согласно существующему бизнес-правилу: подходят ли MCC, по которым у компании был оборот в текущем месяце `month`, для подключения продукта `P3`
- `is_relevant_mcc_altpay5` — согласно существующему бизнес-правилу: подходят ли MCC, по которым у компании был оборот в текущем месяце `month`, для подключения платёжного метода `altpay5`
- `is_relevant_turnover_altpay5` — согласно существующему бизнес-правилу: подходит ли оборот компании в текущем месяце `month` для подключения платёжного метода `altpay5`

### Количественные переменные
#### Непрерывные
- `cnt_trx` — кол-во транзакций компании в текущем месяце `month`. Признак логарифмирован и масштабирован
- `turnover` — оборот компании в текущем месяце `month`. Признак логарифмирован и масштабирован
- `altpayN_turnover` — оборот по платёжному методу `N` в текущем месяце `month`. Признак логарифмирован и масштабирован
- `avg_check` — средний чек компании в текущем месяце `month`. Признак логарифмирован и масштабирован
- `turnover_ma_Wm` — скользящий средний оборот компании с окном `W` на момент текущего месяца `month`. Признаки логарифмированы и масштабированы
- `turnover_ema_Wm` — экспоненциальный скользящий средний оборот компании с окном `W` на момент текущего месяца `month`. Признаки логарифмированы и масштабированы
- `cnt_trx_ma_Wm` — скользящее среднее кол-во транзакций компании с окном `W` на момент текущего месяца `month`. Признаки логарифмированы и масштабированы
- `cnt_trx_ema_Wm` — экспоненциальное скользящее среднее кол-во транзакций компании с окном `W` на момент текущего месяца `month`. Признаки логарифмированы и масштабированы
- `avg_check_wma_Wm` — взвешенный скользящий средний чек компании с окном `W` на момент текущего месяца `month`. Признаки логарифмированы и масштабированы
- `turnover_cv_Wm` — скользящий коэффициент вариации оборота компании с окном `W` на момент текущего месяца `month`
- `turnover_change_to_prev_month` — изменение оборота компании в текущем месяце `month` по отношению к предыдущему месяцу `month - 1`
- `cnt_trx_change_to_prev_month` — изменение кол-ва транзакций компании в текущем месяце `month` по отношению к предыдущему месяцу `month - 1`
- `avg_check_change_to_prev_month` — изменение среднего чека компании в текущем месяце `month` по отношению к предыдущему месяцу `month - 1`
- `turnover_change_to_turnover_ma_Wm` — изменение оборота компании в текущем месяце `month` по отношению к скользящему среднему обороту с окном `W`
- `turnover_change_to_turnover_ema_Wm` — изменение оборота компании в текущем месяце `month` по отношению к экспоненциальному скользящему среднему обороту с окном `W`
- `cnt_trx_change_to_cnt_trx_ma_Wm` — изменение кол-ва транзакций компании в текущем месяце `month` по отношению к скользящему среднему кол-ву транзакций с окном `W`
- `cnt_trx_change_to_cnt_trx_ema_Wm` — изменение кол-ва транзакций компании в текущем месяце `month` по отношению к экспоненциальному скользящему среднему кол-ву транзакций с окном `W`
- `avg_check_change_to_avg_check_wma_Wm` — изменение кол-ва транзакций компании в текущем месяце `month` по отношению к скользящему среднему кол-ву транзакций с окном `W`
#### Дискретные
- `lifetime_month_streak_inn` — кол-во активных месяцев жизни подряд на момент текущего месяца `month`
- `active_product_cnt` — кол-во активных продуктов на момент текущего месяца `month`. Не включает продукты, являющиеся таргетами
- `products_ever_connected` — кол-во подключённых продуктов за всё время на момент текущего месяца `month`. В отличие от `active_product_cnt`, включает продукты, являющиеся таргетами
- `active_mcc_cnt` — кол-во MCC, по которым у компании были транзакции в текущем месяце
- `active_gateway_cnt` — кол-во шлюзов, по которым у компании были транзакции в текущем месяце

## Обработка пропусков

Посмотрим общую информацию о датасете.

In [3]:
base_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 390535 entries, 0 to 390534
Data columns (total 66 columns):
 #   Column                                Non-Null Count   Dtype  
---  ------                                --------------   -----  
 0   month                                 390535 non-null  str    
 1   inn                                   390535 non-null  str    
 2   inn_status                            390535 non-null  str    
 3   top_mcc_group_inn                     238570 non-null  str    
 4   who_is_this_first_inn                 390520 non-null  str    
 5   first_trx_month_inn                   390535 non-null  str    
 6   last_active_month_inn                 239559 non-null  str    
 7   lifetime_month_streak_inn             390535 non-null  int64  
 8   real_kam_on_inn                       390535 non-null  bool   
 9   has_termination_start                 390535 non-null  bool   
 10  altpay1_turnover                      390535 non-null  float64
 11  altpay2_tur

Исходный датасет содержит 390535 строк и 67 столбцов. Приведём столбцы с датами и bool к корректному типу данных.

In [4]:
from acquiring_xsell.preprocessing import process_dtypes

base_df = process_dtypes(base_df)

Оценим пропуски в датасете: выведем их кол-во и долю от всего датафрейма.

In [5]:
df_isna = pd.DataFrame({
    'dtype': base_df.dtypes,
    'notna_cnt': base_df.notna().sum(),
    'na_cnt': base_df.isna().sum(),
    'share_of_na': round(base_df.isna().sum() / len(base_df), 2)
})

df_isna = df_isna.query("share_of_na > 0").sort_values(by='share_of_na', ascending=False)

df_isna

,dtype,notna_cnt,na_cnt,share_of_na
first_month_p3,datetime64[us],2043,388492,0.99
other_active_product_cnt,float64,10485,380050,0.97
first_month_altpay5,datetime64[us],11333,379202,0.97
first_month_p2,datetime64[us],25521,365014,0.93
first_month_p1,datetime64[us],35343,355192,0.91
top_mcc_group_inn,str,238570,151965,0.39
last_active_month_inn,datetime64[us],239559,150976,0.39
turnover_cv_2m,float64,362022,28513,0.07
turnover_cv_3m,float64,362022,28513,0.07
turnover_cv_6m,float64,362022,28513,0.07


Наблюдаем признаки с более чем половиной пропущенных значений. Среди них особенно выделяются первые месяцы, когда были подключены дополнительные продукты, доля строк с пропусками в этих столбцах составляет 91–99%, но т.к. это столбцы, из которых в дальнейшем будут составлены целевые переменные, обработку пропусков в них отложим. Начнём с категориального признака `top_mcc_group_inn`. Предположим, что он не принимает никакого значения, если компания на момент текущего месяца находится в оттоке. Нет транзакций <=> нет оборота => нет MCC => нет группы MCC. Проверим это предположение.

In [6]:
base_df.query("top_mcc_group_inn.isna()").groupby('inn_status').size().sort_values(ascending=False)

inn_status
Churn        151597
Active          298
New Churn        55
Reborn           15
dtype: int64

Подавляющее большинство строк с пропусками в `top_mcc_group_inn` действительно имеют `inn_status == 'Churn'`, однако есть кейсы, когда компания является активной (т.е. имеет завершённые транзакции в отчётном месяце), но всё равно не имеет группу MCC. Такого физически не может быть в интернет-эквайринге, т.к. каждая транзакция имеет свой MCC. Возможно, это ошибка в существующей классификации групп MCC, и на самом деле MCC есть, но он не попал ни в какую группу, поэтому значение `NaN`. Заполним пропуски для оттёкших и активных клиентов разными значениями: для оттёкших заполним пропуски значением `'Churn'`, а для активных — `'No MCC Group'`.

In [7]:
base_df.loc[
    (base_df['inn_status'].isin(['Churn', 'New Churn'])) & (base_df['top_mcc_group_inn'].isna()),
    'top_mcc_group_inn'
] = 'Churn'

base_df.loc[
    base_df['inn_status'].isin(['Active', 'Reborn']) & (base_df['top_mcc_group_inn'].isna()),
    'top_mcc_group_inn'
] = 'No MCC Group'

Перейдём к обработке пропусков в `datetime`-признаках.

In [8]:
df_isna[df_isna['dtype'] == 'datetime64[us]']

,dtype,notna_cnt,na_cnt,share_of_na
first_month_p3,datetime64[us],2043,388492,0.99
first_month_altpay5,datetime64[us],11333,379202,0.97
first_month_p2,datetime64[us],25521,365014,0.93
first_month_p1,datetime64[us],35343,355192,0.91
last_active_month_inn,datetime64[us],239559,150976,0.39


В столбцах вида `first_month_Y` содержатся даты подключения дополнительных продуктов, эти столбцы будут использованы в дальнейшем для определения целевых переменных, сейчас их пропускаем. Однако столбец `last_active_month_inn` к ним не относится, попробуем заполнить пропуски в нём.

In [9]:
base_df[base_df['last_active_month_inn'].isna()].groupby('inn_status').size()

inn_status
Churn    150976
dtype: int64

Видим, что все без исключения строки, в которых отсутствует `last_active_month` имеют `inn_status == 'Churn'`. Причём что примечательно, нет ни одного наблюдения в статусе `'New Churn'`. Возможно, это клиенты, оттёкшие давно и не имеющие активного месяца за период, в рамках которого были использованы данные о транзакциях в процессе сборки датасета. Оценим кол-во таких компаний по годам привлечения.

In [10]:
base_df[
    (base_df['last_active_month_inn'].isna())
].groupby(base_df['first_trx_month_inn'].dt.year)['inn'].nunique()

first_trx_month_inn
2018    1254
2019    1547
2020    3222
2021    3047
2022     939
2023     151
Name: inn, dtype: int64

Действительно, видим что пропуски в `last_active_month` встречаются у старого привлечения. Заполним столбец константным значением — минимальный `last_active_month_inn` по всему датасету - 2 месяца. Возможно, транзакционные данные действительно были использованы только с какого-то определённого года, поэтому последний месяц активности, если у ИНН он был до начала этого периода, в датасете не определён. Берём разницу в 2 месяца, а не в 1, т.к. если бы мы взяли 1 месяц, необходимо было бы проставить другой `inn_status` — `'New Churn'`. На текущем этапе разумным решением будет не усложнять логику, однако в дальнейшем её можно будет усовершенствовать, т.к. по этому константному значению мы точно сможем определить строки, в которых они были проставлены.

In [11]:
min_last_active_month = base_df['last_active_month_inn'].min()
base_df['last_active_month_inn'] = base_df['last_active_month_inn'].fillna(min_last_active_month - pd.DateOffset(months=2))

Перейдём к заполнению пропусков в числовых признаках.

In [12]:
df_isna[df_isna['dtype'] == 'float64']

,dtype,notna_cnt,na_cnt,share_of_na
other_active_product_cnt,float64,10485,380050,0.97
turnover_cv_2m,float64,362022,28513,0.07
turnover_cv_3m,float64,362022,28513,0.07
turnover_cv_6m,float64,362022,28513,0.07
turnover_change_to_prev_month,float64,362022,28513,0.07
cnt_trx_change_to_prev_month,float64,362022,28513,0.07
avg_check_change_to_prev_month,float64,362022,28513,0.07


Пропуски в `other_active_product_cnt` заполним нулями. Если нет данных об активных продуктах на какой-либо месяц, считаем что активности не было.

In [13]:
base_df['other_active_product_cnt'] = base_df['other_active_product_cnt'].fillna(0)

Перейдём к метрикам отношения к предыдущему месяцу. Возможно, из-за того что метрики отношения к предыдущему месяцу нельзя рассчитать, пустые значения проставлены в строках где `month == first_trx_month_inn` или `month == '2024-12-01'`. Проверим обратное утверждение: есть ли строки, которые не соответствуют условиям выше, но всё равно имеют пустые значения в столбцах вида `*_change_to_prev_month`?

In [14]:
change_to_prev_month_cols = [col for col in base_df.columns if col.endswith("_change_to_prev_month")]

mask_na = (
    (base_df["month"] == base_df["first_trx_month_inn"])
    | (base_df["month"] == base_df["month"].min())
)

base_df[~mask_na & base_df[change_to_prev_month_cols].min(axis=1).isna()].shape[0]

0

Гипотеза подтвердилась. Заполним пропуски нулями и добавим флаги `is_first_trx_month_inn` и `is_left_time_border` — они будут соответственно указывать, что у ИНН текущий месяц соответствует месяцу первой транзакции, и что `month` в этой строке является минимальный по всему датасету, из-за чего мы не можем корректно рассчитать метрики отношения к предыдущему месяцу.

In [15]:
base_df[change_to_prev_month_cols] = base_df[change_to_prev_month_cols].fillna(0)

base_df['is_first_trx_month'] = base_df["month"] == base_df["first_trx_month_inn"]
base_df['is_left_time_border'] = base_df["month"] == base_df["month"].min()

Проверим ту же гипотезу на скользящих коэффициентах вариации оборота.

In [16]:
cv_cols = [col for col in base_df.columns if "_cv_" in col]

mask_na = (
    (base_df["month"] == base_df["first_trx_month_inn"])
    | (base_df["month"] == base_df["month"].min())
)

base_df[~mask_na & base_df[cv_cols].min(axis=1).isna()].shape[0]

0

Ожидаемо, ситуация такая же: т.к. коэффициент вариации рассчитывается как `std / mean`, а стандартное отклонение невозможно рассчитать менее чем по двум наблюдениям, соответственно по первым месяцам жизни ИНН и минимальному месяцу в датасете CV принимает значение nan. Обработаем его также как метрики отношения к предыдущему месяцу. Новые флаги не добавляем, т.к. уже рассчитали `is_first_trx_month_inn` и `is_left_time_border` выше.

In [17]:
base_df[cv_cols] = base_df[cv_cols].fillna(0)

Остался последний признак с пропущенными значениями: `who_is_this_first_inn`. Пропущенных значений всего 15, посмотрим, что это за наблюдения.

In [18]:
base_df[base_df['who_is_this_first_inn'].isna()]

,month,inn,inn_status,top_mcc_group_inn,who_is_this_first_inn,first_trx_month_inn,last_active_month_inn,lifetime_month_streak_inn,real_kam_on_inn,has_termination_start,altpay1_turnover,altpay2_turnover,altpay3_turnover,altpay4_turnover,altpay5_turnover,altpay6_turnover,has_recurrent_payments,cnt_trx,turnover,other_active_product_cnt,is_relevant_mcc_p3,is_relevant_mcc_altpay5,is_relevant_turnover_altpay5,active_mcc_cnt,active_gateway_cnt,avg_check,turnover_ma_2m,turnover_ema_2m,turnover_ma_3m,turnover_ema_3m,turnover_ma_6m,turnover_ema_6m,cnt_trx_ma_2m,cnt_trx_ema_2m,cnt_trx_ma_3m,cnt_trx_ema_3m,cnt_trx_ma_6m,cnt_trx_ema_6m,avg_check_wma_2m,avg_check_wma_3m,avg_check_wma_6m,turnover_cv_2m,turnover_cv_3m,turnover_cv_6m,turnover_change_to_prev_month,cnt_trx_change_to_prev_month,avg_check_change_to_prev_month,turnover_change_to_turnover_ma_2m,turnover_change_to_turnover_ma_3m,turnover_change_to_turnover_ma_6m,turnover_change_to_turnover_ema_2m,turnover_change_to_turnover_ema_3m,turnover_change_to_turnover_ema_6m,cnt_trx_change_to_cnt_trx_ma_2m,cnt_trx_change_to_cnt_trx_ma_3m,cnt_trx_change_to_cnt_trx_ma_6m,cnt_trx_change_to_cnt_trx_ema_2m,cnt_trx_change_to_cnt_trx_ema_3m,cnt_trx_change_to_cnt_trx_ema_6m,avg_check_change_to_avg_check_wma_2m,avg_check_change_to_avg_check_wma_3m,avg_check_change_to_avg_check_wma_6m,first_month_p1,first_month_p2,first_month_p3,first_month_altpay5,is_first_trx_month,is_left_time_border
49743,2025-04-01,INN_8892,Churn,Churn,NaN,2018-02-01,2022-11-01,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0,0,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaT,NaT,NaT,NaT,False,False
90840,2024-12-01,INN_8892,Churn,Churn,NaN,2018-02-01,2022-11-01,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0,0,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaT,NaT,NaT,NaT,False,True
90841,2025-01-01,INN_8892,Churn,Churn,NaN,2018-02-01,2022-11-01,0,1,0,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0,0,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaT,NaT,NaT,NaT,False,False
98108,2025-02-01,INN_8892,Churn,Churn,NaN,2018-02-01,2022-11-01,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0,0,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaT,NaT,NaT,NaT,False,False
98109,2025-03-01,INN_8892,Churn,Churn,NaN,2018-02-01,2022-11-01,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0,0,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaT,NaT,NaT,NaT,False,False
103888,2025-05-01,INN_8892,Churn,Churn,NaN,2018-02-01,2022-11-01,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0,0,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaT,NaT,NaT,NaT,False,False
103889,2026-01-01,INN_8892,Churn,Churn,NaN,2018-02-01,2022-11-01,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0,0,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaT,NaT,NaT,NaT,False,False
157358,2025-10-01,INN_8892,Churn,Churn,NaN,2018-02-01,2022-11-01,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0,0,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaT,NaT,NaT,NaT,False,False
157359,2025-11-01,INN_8892,Churn,Churn,NaN,2018-02-01,2022-11-01,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0,0,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0

Пропущенные значения встречаются только у одной ИНН в статусе отток. Заметим, что по всем месяцам жизни у этой ИНН изменился только один признак — `real_kam_on_inn`. При этом ни одного из продуктов у неё нет. Удалим эту ИНН из данных, чтобы не создавать ещё одну категорию `who_is_this_first_inn`

In [19]:
base_df = base_df[base_df['who_is_this_first_inn'].notna()]

Проверим, остались ли столбцы с пропусками, кроме таргетных.

In [20]:
base_df.isna().sum().sort_values(ascending=False)

first_month_p3         388477
first_month_altpay5    379187
first_month_p2         364999
first_month_p1         355177
month                       0
                        ...  
turnover_ma_3m              0
turnover_ema_3m             0
turnover_ma_6m              0
turnover_ema_6m             0
is_left_time_border         0
Length: 68, dtype: int64

Кроме столбцов `first_month_Y`, на которых будут рассчитаны таргетные переменные, пропусков не осталось. Переходим к определению таргетов.

## Определение таргетов и разделение датасетов

В рамках данной работы таргетом будет являться факт подключения нового продукта у ИНН в следующем месяце. Соответственно, для каждого продукта необходимо собрать такой датасет, в котором по каждой ИНН будут только месяцы жизни строго до месяца подключения нового продукта. Продукты `'p1'` и `'p2'` по бизнес-правилу взаимоисключающие, поэтому в рамках данной работы для них будет собран один общий датасет. В нём таргетом будем считать по каждой ИНН строку с месяцем, предыдущим перед месяцем подключения первого продукта из множества `('p1', 'p2')`.

In [21]:
def make_product_df(base_df: pd.DataFrame, product: str) -> pd.DataFrame:
    """Собирает датафрейм для переданного продукта и определяет таргет как последний месяц перед подключением"""
    df = base_df.copy()

    first_month_col = f'first_month_{product}'

    # Оставляем только строки до подключения продукта
    df = df[
        (df['month'] < df[first_month_col]) | (df[first_month_col].isna())
    ].reset_index(drop=True)

    df['target'] = (
        df['month'] + pd.DateOffset(months=1) == df[first_month_col]
    ).astype('int8')

    df = df.drop(columns=[first_month_col])

    # Устраняем Data Leakage по другим продуктам
    other_product_cols = [col for col in df.columns if col.startswith("first_month_")]
    for col in other_product_cols:
        df[col] = df[
            (df['month'] >= df[col]) | df[col].isna()
        ][col]

    return df


# Определяем общий таргет для продуктов p1 и p2
base_df['first_month_p1p2'] = base_df[['first_month_p1', 'first_month_p2']].min(axis=1)
base_df = base_df.drop(columns=['first_month_p1', 'first_month_p2'])

df_p1p2 = make_product_df(base_df=base_df, product='p1p2')
df_p3 = make_product_df(base_df=base_df, product='p3')
df_altpay5 = make_product_df(base_df=base_df, product='altpay5')

dfs = [df_p1p2, df_p3, df_altpay5]

Оценим распределение классов в датафреймах по каждому продукту.

In [22]:
print("=== Баланс классов ===")
print("Всего наблюдений | Положительный класс | Доля положительного класса")
for df in dfs:
    share_of_positive = df.target.sum() / len(df)
    print(f"{len(df):<16} | {int(df.target.sum()):<19} | {share_of_positive:.2%}")

=== Баланс классов ===
Всего наблюдений | Положительный класс | Доля положительного класса
336921           | 776                 | 0.23%
388615           | 26                  | 0.01%
381734           | 512                 | 0.13%


Наблюдаем экстремальный дисбаланс классов. Это ожидаемо, т.к. в исходном датасете собраны данные по месяцам жизни клиентов компании, привлечённых за всё время. Попробуем увеличить долю положительного класса за счёт удаления строк, где `inn_status in ('Churn', 'New Churn')`. Но сначала проверим, встречаются ли кейсы, когда клиент в статусе «Отток» по основному продукту (интернет-эквайринг) в следующем месяце подключил новый доп. продукт.

In [23]:
print("=== Положительный класс ===")
print("Всего | В оттоке | Доля")
for df in dfs:
    positive_in_churn = df.query("inn_status in ('Churn', 'New Churn')").target.sum()
    share_of_positive_in_churn = positive_in_churn / df.target.sum()
    print(f"{df.target.sum():<5} | {int(positive_in_churn):<8} | {share_of_positive_in_churn:.2%}")

=== Положительный класс ===
Всего | В оттоке | Доля
776   | 311      | 40.08%
26    | 9        | 34.62%
512   | 104      | 20.31%


Такие кейсы встречаются, их доля составляет 20–40% от общего кол-ва положительных классов. В силу редкости объектов с положительным таргетом, их количественное уменьшение для нас крайне нежелательно. Однако, всё же проверим, как изменятся доли положительного класса в датасетах, если мы решим избавиться от строк с условием `inn_status in ('Churn', 'New Churn')`.

In [24]:
print("=== Баланс классов (только активные ИНН) ===")
print("Всего наблюдений | Положительный класс | Доля положительного класса")
for df in dfs:
    df_active = df.query("inn_status not in ('Churn', 'New Churn')").copy()
    share_of_positive = df_active.target.sum() / len(df_active)
    print(f"{len(df_active):<16} | {int(df_active.target.sum()):<19} | {share_of_positive:.2%}")

=== Баланс классов (только активные ИНН) ===
Всего наблюдений | Положительный класс | Доля положительного класса
81717            | 465                 | 0.57%
113213           | 17                  | 0.02%
107209           | 408                 | 0.38%


Доля положительного класса увеличилась в 2–3 раза, однако значения по-прежнему экстремально малые. Примем это к сведению, экспортируем датасеты по продуктам с заполнеными пропусками и приступим непосредственно к EDA по каждому продукту.

In [ ]:
from pathlib import Path

path_p1p2 = Path("..", "data", "processed", "p1p2_data.csv")
path_p3 = Path("..", "data", "processed", "p3_data.csv")
path_altpay5 = Path("..", "data", "processed", "altpay5_data.csv")

df_p1p2.to_csv(path_p1p2, index=False, sep=";")
df_p3.to_csv(path_p3, index=False, sep=";")
df_altpay5.to_csv(path_altpay5, index=False, sep=";")